src/sync/sync_user.cpp

```diff
@@ -0,0 +1,166 @@
+    if (m_state == State::Error) {
+        return sessions;
+    }
+    for (auto it : m_sessions) {
+        if (auto ptr = it.second.lock()) {
+            if (ptr->is_valid()) {
+                sessions.emplace_back(std::move(ptr));
+                continue;
+            }
+        }
+        keys_to_remove.emplace_back(it.first);
+    }
+    for (auto key : keys_to_remove) {
+        // Clear any sessions that have been destroyed or fatally errored out.
+        m_sessions.erase(key);
+    }
+    return sessions;
+}
+
+std::shared_ptr<SyncSession> SyncUser::session_for_url(const std::string& url)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_state == State::Error) {
+        return nullptr;
+    }
+    auto it = m_sessions.find(url);
+    if (it == m_sessions.end()) {
+        return nullptr;
+    }
+    auto locked = it->second.lock();
+    if (!locked) {
+        // Remove the session from the map, because it has fatally errored out or the entry is invalid.
+        m_sessions.erase(it);
+    }
+    return locked;
+}
+
+void SyncUser::update_refresh_token(std::string token)
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+    switch (m_state) {
+        case State::Error:
+            return;
+        case State::Active:
+            m_refresh_token = std::move(token);
+            break;
+        case State::LoggedOut: {
+            m_refresh_token = std::move(token);
+            m_state = State::Active;
+            // (Re)activate all pending sessions.
+            for (auto pair : m_waiting_sessions) {
+                if (auto ptr = pair.second.lock()) {
+                    m_sessions[pair.first] = ptr;
+                    lock.unlock();
+                    ptr->revive_if_needed();
+                }
+            }
+            m_waiting_sessions.clear();
+            break;
+        }
+    }
+    if (auto manager = SyncManager::shared().metadata_manager()) {
+        // Update persistent user metadata.
+        auto metadata = SyncUserMetadata(*manager, m_identity);
+        metadata.set_state(server_url, token);
+    }
+}
+
+void SyncUser::log_out()
+{
+    if (m_is_admin) {
+        // Admin users cannot be logged out.
+        return;
+    }
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_state == State::LoggedOut) {
+        return;
+    }
+    m_state = State::LoggedOut;
+    // Move all active sessions into the waiting sessions pool. If the user is
+    // logged back in, they will automatically be reactivated.
+    for (auto pair : m_sessions) {
+        if (auto ptr = pair.second.lock()) {
+            ptr->log_out();
+            m_waiting_sessions[pair.first] = ptr;
+        }
+    }
+    m_sessions.clear();
+    if (auto manager = SyncManager::shared().metadata_manager()) {
+        // Mark the user as 'dead' in the persisted metadata Realm.
+        auto metadata = SyncUserMetadata(*manager, m_identity, false);
+        metadata.mark_for_removal();
```

Comment:
`metadata` is basically a barebones Realm object, and `mark_for_removal()` just sets a property on it. If you want I'm happy to sit down with you at our upcoming shindig and explain the metadata (and other) code I wrote to you in person.